### PR TITLE
contrib: More restrictive dcrd service privileges.

### PR DESCRIPTION
**Pull request by zhizhongzhiwai**
more restrict dcrd privilege

### CODE REVIEW

File: contrib/services/systemd/dcrd.service [hunk 1]
```diff
@@ -1,13 +1,53 @@
 [Unit]
 Description=Decred Full Node
+Requires=local-fs.target network-online.target network.target
+After=local-fs.target network-online.target network.target
 
 [Service]
 Type=simple
+
+# modify to your username
```
Comment (davecgh) [inline contrib/services/systemd/dcrd.service:9]: This is intentionally not the user's account.  It is expected that the service will be run under a separate account and group which is best practice for privilege separation.

File: contrib/services/systemd/dcrd.service [hunk 2]
```diff
@@ -1,13 +1,53 @@
...
+ReadWritePaths=%h/.dcrd/
+
+# dcrd datadir, modify if data directory on other mount device. e.g /data/dcrd-data/
+ReadWritePaths=%h/.dcrd/data
+
+# dcrd logdir
+ReadWritePaths=%h/.dcrd/logs
+
+# launch dcrd
+ExecStart=%h/decred/dcrd
```
Comment (davecgh) [inline contrib/services/systemd/dcrd.service:23]: Again, this (and thus the others) were intentionally not the user's home directory.  System-wide services are expected in certain locations and _not_ the user's home directory.

**Review by davecgh — CHANGES_REQUESTED** [review]
Thanks for the PR.  The additional strict privileges are a nice addition!  There are some changes that are not best practices though too and I've identified them with inline comments.

File: contrib/services/systemd/dcrd.service [hunk 3]
```diff
@@ -27,6 +27,7 @@ Restart=on-abnormal
 RestartSec=10s
 
 # strict privilege
+InaccessiblePaths=/usr/local/sbin /usr/local/bin /usr/sbin /usr/bin /sbin /bin
```
Comment (jrick) [inline contrib/services/systemd/dcrd.service:30]: restricting read access to /usr and/or /usr/local, where Go is typically installed from packages, may result in missing information in stack traces

Comment (jrick) [reply contrib/services/systemd/dcrd.service:30]: nvm, all of this should be baked into the executable.

**Comment by davecgh** [comment]
I'll be testing this out within a few days.  It reads well, but I want to test it before approving.

**Review by davecgh — COMMENTED** [review]
I tested this with Ubuntu 20.04.1 LTS with a user and group named dcrd.

It seems to work for the most part, but there are a couple of errors that look incorrect to me:

```bash
Sep 29 14:23:09 dcru systemd[1]: Started Decred Full Node.
Sep 29 14:23:12 dcru dcrd[655]: 2020-09-29 14:23:12.585 [INF] DCRD: Version 1.6.0-pre (Go version go1.15.2 linux/amd64)
Sep 29 14:23:13 dcru dcrd[655]: 2020-09-29 14:23:12.594 [INF] DCRD: Home dir: /home/dcrd/.dcrd
Sep 29 14:23:13 dcru dcrd[655]: 2020-09-29 14:23:12.595 [INF] DCRD: Loading block database from '/home/dcrd/.dcrd/data/mainnet/blocks_ffldb'
Sep 29 14:23:13 dcru dcrd[655]: 2020-09-29 14:23:13.376 [INF] DCRD: Block database loaded
Sep 29 14:23:13 dcru dcrd[655]: 2020-09-29 14:23:13.379 [WRN] AMGR: Skipping bound address 0.0.0.0:9108: route ip+net: netlinkrib: address family not supported by protocol
Sep 29 14:23:13 dcru dcrd[655]: 2020-09-29 14:23:13.379 [WRN] AMGR: Skipping bound address [::]:9108: route ip+net: netlinkrib: address family not supported by protocol
Sep 29 14:23:13 dcru dcrd[655]: 2020-09-29 14:23:13.380 [INF] INDX: Exists address index is enabled
Sep 29 14:23:13 dcru dcrd[655]: 2020-09-29 14:23:13.380 [INF] INDX: CF index is enabled
Sep 29 14:23:14 dcru dcrd[655]: 2020-09-29 14:23:14.060 [INF] CHAN: Loading block index...
Sep 29 14:23:20 dcru dcrd[655]: 2020-09-29 14:23:20.965 [INF] STKE: Stake database version 1 loaded
Sep 29 14:23:21 dcru dcrd[655]: 2020-09-29 14:23:21.006 [INF] CHAN: Blockchain database version info: chain: 7, compression: 1, block index: 3
Sep 29 14:23:21 dcru dcrd[655]: 2020-09-29 14:23:21.006 [INF] CHAN: Chain state: height 31066, hash 0000000000001594176ac75df0879e8e8a095fc2e8bd72ee517d736f90509fa0, total transactions 447315, work 50321630581817831043, stake version 0
Sep 29 14:23:21 dcru dcrd[655]: 2020-09-29 14:23:21.169 [INF] AMGR: Loaded 32 addresses from file '/home/dcrd/.dcrd/data/mainnet/peers.json'
Sep 29 14:23:21 dcru dcrd[655]: 2020-09-29 14:23:21.169 [INF] RPCS: RPC server listening on 127.0.0.1:9109
Sep 29 14:23:21 dcru dcrd[655]: 2020-09-29 14:23:21.192 [INF] CMGR: Server listening on 0.0.0.0:9108
Sep 29 14:23:21 dcru dcrd[655]: 2020-09-29 14:23:21.192 [INF] CMGR: Server listening on [::]:9108
...
Sep 29 14:27:39 dcru systemd[1]: Stopping Decred Full Node...
Sep 29 14:27:39 dcru dcrd[655]: 2020-09-29 14:27:39.067 [INF] DCRD: Received signal (terminated).  Shutting down...
Sep 29 14:27:39 dcru dcrd[655]: 2020-09-29 14:27:39.068 [WRN] SRVR: Server shutting down
Sep 29 14:27:39 dcru dcrd[655]: 2020-09-29 14:27:39.068 [WRN] RPCS: RPC server shutting down
Sep 29 14:27:39 dcru dcrd[655]: 2020-09-29 14:27:39.068 [INF] RPCS: RPC server shutdown complete
Sep 29 14:27:39 dcru dcrd[655]: 2020-09-29 14:27:39.069 [INF] BMGR: Block manager shutting down
Sep 29 14:27:48 dcru dcrd[655]: 2020-09-29 14:27:48.925 [INF] BMGR: Processed 726 blocks in the last 20.75s (4183 transactions, 3408 tickets, 3557 votes, 62 revocations, height 53929, 2016-08-12 16:56:35 -0500 CDT)
Sep 29 14:27:49 dcru dcrd[655]: 2020-09-29 14:27:48.926 [INF] AMGR: Address manager shutting down
Sep 29 14:27:49 dcru dcrd[655]: 2020-09-29 14:27:48.927 [INF] SRVR: Server shutdown complete
Sep 29 14:27:49 dcru dcrd[655]: 2020-09-29 14:27:48.927 [INF] DCRD: Gracefully shutting down the database...
Sep 29 14:27:49 dcru systemd[1]: dcrd.service: State 'stop-sigterm' timed out. Killing.
Sep 29 14:27:49 dcru systemd[1]: dcrd.service: Killing process 655 (dcrd) with signal SIGKILL.
Sep 29 14:27:49 dcru systemd[1]: dcrd.service: Killing process 694 (dcrd) with signal SIGKILL.
Sep 29 14:27:49 dcru systemd[1]: dcrd.service: Killing process 695 (dcrd) with signal SIGKILL.
Sep 29 14:27:49 dcru systemd[1]: dcrd.service: Killing process 696 (dcrd) with signal SIGKILL.
Sep 29 14:27:49 dcru systemd[1]: dcrd.service: Killing process 697 (dcrd) with signal SIGKILL.
Sep 29 14:27:49 dcru systemd[1]: dcrd.service: Killing process 709 (dcrd) with signal SIGKILL.
Sep 29 14:27:49 dcru systemd[1]: dcrd.service: Killing process 710 (dcrd) with signal SIGKILL.
Sep 29 14:27:49 dcru systemd[1]: dcrd.service: Killing process 2135 (dcrd) with signal SIGKILL.
Sep 29 14:27:49 dcru systemd[1]: dcrd.service: Killing process 2154 (dcrd) with signal SIGKILL.
Sep 29 14:27:49 dcru systemd[1]: dcrd.service: Main process exited, code=killed, status=9/KILL
Sep 29 14:27:49 dcru systemd[1]: dcrd.service: Failed with result 'timeout'.
Sep 29 14:27:49 dcru systemd[1]: Stopped Decred Full Node.
```

Notice the:

```
[WRN] AMGR: Skipping bound address 0.0.0.0:9108: route ip+net: netlinkrib: address family not supported by protocol
[WRN] AMGR: Skipping bound address [::]:9108: route ip+net: netlinkrib: address family not supported by protocol
```

**Comment by davecgh** [comment]
Following up, removing the `RestrictAddressFamilies` line resolves the issue.

File: contrib/services/systemd/dcrd.service [hunk 4]
```diff
@@ -1,13 +1,53 @@
...
+PrivateTmp=true
+PrivateUsers=true
+ProtectClock=true
+ProtectControlGroups=true
+ProtectHostname=true
+ProtectKernelModules=true
+ProtectKernelTunables=true
+ProtectSystem=strict
+RemoveIPC=true
+RestrictAddressFamilies=AF_INET AF_INET6 AF_UNIX
```
Comment (davecgh) [inline contrib/services/systemd/dcrd.service:45]: As discussed in the comments, this needs to be removed to prevent address manager issues.

**Review by davecgh — CHANGES_REQUESTED** [review]

**Comment by zhizhongzhiwai** [comment]
Apologize for the inconvenience。 i ignored the WARNING.
the WARNING caused by miss AF_NETLINK protocol family. fixed it.  
after add AF_NETLINK dcrd 1.5.2 run well with ubuntu:20.04 and debian 10.

**Review by davecgh — APPROVED** [review]
Tested with the latest update and there are no more warnings and I was able to fully sync the chain and interact with dcrd via the JSON-RPC server without issue.

**Pull request merged**